src/transfers.cpp

```diff
@@ -36,31 +36,78 @@ Rcpp::List rcpp_transfer_nbs (Rcpp::DataFrame stops,
     const std::vector <double> stop_x = stops ["stop_lon"];
     const std::vector <double> stop_y = stops ["stop_lat"];
 
+    std::unordered_map <std::pair <double, double>, std::unordered_set <size_t>, pair_hash> stop_map;
+    std::unordered_map <size_t, std::pair <double, double>> stop_index;
+    size_t count = 0;
+    for (size_t i = 0; i < n; i++)
+    {
+        std::pair <double, double> xy_pair {stop_x [i], stop_y [i]};
+
+        std::unordered_set <size_t> index_set;
+        if (stop_map.find (xy_pair) != stop_map.end ())
+        {
+            index_set = stop_map.at (xy_pair);
+        } else
+        {
+            stop_index.emplace (count++, xy_pair);
+        }
+        index_set.emplace (i);
+
+        stop_map.erase (xy_pair);
+        stop_map.emplace (xy_pair, index_set);
+    }
+
+    const size_t n_unique = stop_map.size ();
+
     Rcpp::List res (n * 2);
 
-    for (size_t i = 0; i < n; i++)
+    for (size_t i = 0; i < n_unique; i++)
     {
-        const double cosy1 = cos (stop_y [i] * pi / 180.0);
+        std::pair <double, double> xy_i = stop_index.at (i);
+        std::unordered_set <size_t> index_i = stop_map.at (xy_i);
+
+        const double cosy1 = cos (xy_i.second * pi / 180.0);
 
         std::vector <size_t> index;
         std::vector <double> dist;
-        for (size_t j = 0; j < n; j++)
```

Comment:
It seems to me that you implemented a different optimization from what i was originally thinking of 😅 My naive assumption was that basically, you could just change this line to `for (size_t j = i; j < n; j++)` and thereby skip calculations that were already done in the other direction.

Reply:
Oh yeah, right; of course! I'll implement that. It's all a bit tricky now, because the `i` and `j` values are no longer direct indices, so can't be done with a `j = i` approach, but i'll figure out a way to index it back and do that. Thanks!